File: Array/yangliu/Solution_06_16.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

public class Solution_06_16 {

    public boolean isValid(String s) {
        char[] chars = s.toCharArray();
        if (chars.length % 2 != 0) return false;
        Stack<Character> characters = new Stack<Character>();
        for (int i = 0; i < chars.length; i++) {
            if (chars[i] == '{' || chars[i] == '(' || chars[i] == '[')
                characters.push(chars[i]);
            else {
                if (!characters.empty())
                    return false;

                Character peek = characters.peek();
                if (peek == '[' && chars[i] == ']' ||
                        peek == '(' && chars[i] == ')' ||
                        peek == '{' && chars[i] == '}')
                    characters.pop();
            }


        }
        return characters.empty();
    }


}

File: Array/yangliu/Solution_06_16.c

int main() {

    return 0;
}

bool isValid(char * s){
    int len = strlen(s);
    if (len == 0)return true;
    if (len %2!=0) return false;
    char stack[len];
    int stackIndex = -1;
    for (int i = 0; i < len; i++) {
        if (s[i] == '(' || s[i] == '[' || s[i] == '{')
            stack[++stackIndex] = s[i];
        else {
            if (stackIndex == -1) { return false; }
            else {
                if (stack[stackIndex] =='('&& s[i] == ')' || stack[stackIndex] =='['&& s[i] == ']' ||
                    stack[stackIndex] =='{'&& s[i] == '}') {
                    --stackIndex;
                }
            }
        }
    }
    return stackIndex == -1;
}
